src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractContinuousIntegrationService.java

```diff
@@ -69,7 +75,61 @@ public Result createResultFromBuildResult(AbstractBuildResultNotificationDTO bui
      * @param result      the result for which the feedback should be added
      * @param buildResult The build result
      */
-    protected abstract void addFeedbackToResult(Result result, AbstractBuildResultNotificationDTO buildResult);
+    private void addFeedbackToResult(Result result, AbstractBuildResultNotificationDTO buildResult) {
+        final var jobs = buildResult.getBuildJobs();
+        final var programmingExercise = (ProgrammingExercise) result.getParticipation().getExercise();
+
+        // 1) add feedback for failed and passed test cases
+        addTestCaseFeedbacksToResult(result, jobs, programmingExercise);
+
+        // 2) process static code analysis feedback
+        addStaticCodeAnalysisFeedbackToResult(result, buildResult, programmingExercise);
+
+        // 3) process testwise coverage analysis report
+        addTestwiseCoverageReportToResult(result, buildResult, programmingExercise);
+
+        // Relevant feedback is negative
+        result.setHasFeedback(result.getFeedbacks().stream().anyMatch(feedback -> !feedback.isPositive() || feedback.getDetailText() != null));
```

Comment:
In `ProgrammingExerciseGradingService` line 672
https://github.com/ls1intum/Artemis/blob/f861347518e77cdf2b2e6ccc41bace48db9ea5a1/src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java#L672
there is an issue with the missing check for `getDetailText()` as well. A fix there would be
```java
if (result.getFeedbacks().stream().allMatch(feedback -> Boolean.TRUE.equals(feedback.isPositive() && feedback.getDetailText() == null)
```

Do you want to add a new method to `Feedback`
```java
    @JsonIgnore
    public boolean hasDetails() {
        return !isPositive() || detailText != null;
    }
```
in this PR to make sure this issue won’t happen in the future, or should I move that to a follow-up PR?

```suggestion
        result.setHasFeedback(result.getFeedbacks().stream().anyMatch(Feedback::hasDetails));
```


In any case, the comment is no longer accurate:
```suggestion
        result.setHasFeedback(result.getFeedbacks().stream().anyMatch(feedback -> !feedback.isPositive() || feedback.getDetailText() != null));
```